Sources/NIO/ControlMessage.swift

```diff
@@ -0,0 +1,224 @@
+    #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
+    fileprivate static let notCapableValue = IPTOS_ECN_NOTECT
+    #else
+    fileprivate static let notCapableValue = IPTOS_ECN_NOT_ECT    // Linux
+    #endif
+    
+    func asCInt() -> CInt {
+        switch self {
+        case .transportNotCapable:
+            return .init(NIOExplicitCongestionNotificationState.notCapableValue)
+        case .transportCapableFlag0:
+            return .init(IPTOS_ECN_ECT0)
+        case .transportCapableFlag1:
+            return .init(IPTOS_ECN_ECT1)
+        case .congestionExperienced:
+            return .init(IPTOS_ECN_CE)
+        }
+    }
+}
+
+struct UnsafeOutboundControlBytes {
+    private var controlBytes: UnsafeMutableRawBufferPointer
+    private var writePosition: size_t = 0
+    
+    /// Control bytes are captured - this structure must not have a lifetime exceeded them.
+    init(controlBytes: UnsafeMutableRawBufferPointer) {
+        self.controlBytes = controlBytes
+    }
+    
+    mutating func appendControlMessage<PayloadType>(level: Int32,
+                                          type: Int32,
+                                          payload: PayloadType) {
+        let writableBuffer = UnsafeMutableRawBufferPointer(rebasing: self.controlBytes[writePosition...])
+        
+        let requiredSize = Posix.cmsgSpace(payloadSize: MemoryLayout.stride(ofValue: payload))
+        precondition(writableBuffer.count >= requiredSize, "Insufficient size for cmsghdr and data")
+        
+        let bufferBase = writableBuffer.baseAddress!
+        let cmsghdrPtr = bufferBase.bindMemory(to: cmsghdr.self, capacity: 1)
+        cmsghdrPtr.pointee.cmsg_level = level
+        cmsghdrPtr.pointee.cmsg_type = type
+        cmsghdrPtr.pointee.cmsg_len = .init(Posix.cmsgLen(payloadSize: MemoryLayout.size(ofValue: payload)))
+        
+        let dataPointer = Posix.cmsgData(for: .some(cmsghdrPtr))
+        precondition(dataPointer!.count >= MemoryLayout<PayloadType>.stride)
+        let dataPointerBase = dataPointer!.baseAddress!
+        let dataPointerTyped = dataPointerBase.bindMemory(to: PayloadType.self, capacity: 1)
+        dataPointerTyped.pointee = payload
+        
+        self.writePosition += requiredSize
+    }
+    
+    /// The result is only valid while this is valid.
+    var validControlBytes: UnsafeMutableRawBufferPointer {
+        if writePosition == 0 {
+            return UnsafeMutableRawBufferPointer(start: nil, count: 0)
+        }
+        return UnsafeMutableRawBufferPointer(rebasing: self.controlBytes[0 ..< self.writePosition])
+    }
+    
+}
+
+extension UnsafeOutboundControlBytes {
+    /// - address:  Either local or remote will do, we just use it for extracting the right protocol.
+    internal mutating func appendExplicitCongestionState(metadata: AddressedEnvelope<ByteBuffer>.Metadata?,
+                                                         address: SocketAddress?) {
```

Comment:
Rather than take the fairly heavy `SocketAddress` struct, can we just take the `NIOBSDSocket.ProtocolFamily` type? That's the question we actually have anyway.